### PR TITLE
Fix/exposed api key

### PR DESCRIPTION
**Pull request by yashpalzala**
PR related to issue: #207 

_currently covered case 2 :_
For old users - we can release a patch wherein we can write a script to extract a user's secret-key that is stored in the
config and put it in secret storage and then delete it from the config.

@wadhia-yash  Raising a PR for single case right now just to know if this changes are good enough and follows all rules as I didn't find much in contributions file.
Please let me know if any suggestions.

### CODE REVIEW

**Comment by samyakkkk** [comment]
Hi @yashpalzala, thanks. Skimmed through it and all looks good. Keep ensuring clean code and comments as you are already doing. 

Look forward to the next iteration.

**Comment by yashpalzala** [comment]
_covered case 1:_
For new users - we already ask for the key in chat view. We need to save it to secret storage now instead of settings.

  commit: [60b0acf](https://github.com/Welltested-AI/fluttergpt/pull/211/commits/60b0acf1041de79fa8feef7e88ea7b94617444e0)

File: src/utilities/secret-storage-manager.ts [hunk 1]
```diff
@@ -24,11 +77,62 @@ export class SecretApiKeyManager{
...
+    }
+    return  await this._context.secrets.delete(this.flutterGptScretStoreKey);
+    
+}
+
+/**
+ * Emits an event only if the apiKey's value is changed. And is disposed off when
+ * extension is disposed.
+ */
+private async onApiKeyChange(): Promise<void> {
```
Comment (samyakkkk) [inline src/utilities/secret-storage-manager.ts]: great to see this case is being handled.

File: src/utilities/exposed-api-key-manager.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            //SET API KEY IN SECRET STORAGE
+           isApiKeyShifted =  await  SecretApiKeyManager.instance.setApiKey(
+                
+                apiKey
+            );
+            
+        }
+        // we delete the key from config only if we successfully shift the key to secret storage
+        if(isApiKeyShifted){
+this.deleteApiKeyFromConfig();
```
Comment (samyakkkk) [inline src/utilities/exposed-api-key-manager.ts]: add a telemetry event please that marks successful shift.

**Comment by samyakkkk** [comment]
@yashpalzala we also made a directory shift recently to move everything in a `/vscode` folder since more platforms are going to be added soon. 

Please take the latest pull at the earliest and we would be available to assist with any merge conflicts. thanks.

File: src/utilities/exposed-api-key-manager.ts (hunk 2 above)
Comment (yashpalzala) [reply src/utilities/exposed-api-key-manager.ts]: cool will add it

**Comment by yashpalzala** [comment]
Hi @samyakkkk 
I have merged the latest changes from develop branch. Have also made changes acc. to new flow for getting api-key. 

Regarding the 3rd case, do you have any specific flow in mind?

**Comment by samyakkkk** [comment]
yes @yashpalzala, let's keep it simple. right by the clear chat button, let's add a profile button. Tapping on the profile button opens up profile edit page with a single option to update the API key. 

please focus majorly on the functionality, the UI could be fixed by us later.  

I'll also request to add tests if possible to ensure that all the three cases are working properly. Best.

**Comment by yashpalzala** [comment]
Hi @samyakkkk I see the branch is not yet merged. Can I help you out with something?

**Comment by samyakkkk** [comment]
Hi @yashpalzala, thanks for checking back. We are not able to ship this yet because we're yet to add the update api key functionality. 

apologies for the delay! we will prioritise it again soon.

**Comment by neiljaywarner** [comment]
@samyakkkk  looking forward to it, thanks @yashpalzala for working on it

**Comment by samyakkkk** [comment]
@yashpalzala I really thank you for your work on this PR! unfortunately we're moving away from the BYOK model and are not asking users their Gemini key anymore. 

I'm sorry we couldn't merge it for this reason.

**Comment by neiljaywarner** [comment]
interesting. what model are you using now, how do you avoid excessive costs
or make back your costs please?

**Comment by samyakkkk** [comment]
We are using a revenue strategy to monetise SDK and API companies who will list on CommandDash. [check plans](https://www.commanddash.io/#pricing) we added today :D

**Comment by neiljaywarner** [comment]
Sounds super fantastic! Thank you 🙂

....sent from my phone

**Pull request closed without merging**